xdsl/frontend/code_generation.py

```diff
@@ -71,6 +72,87 @@ def visit_Assign(self, node: ast.Assign) -> None:
         # TODO: Implement assignemnt in the next patch.
         pass
 
+    def visit_BinOp(self, node: ast.BinOp):
+        op_name: str = node.op.__class__.__name__
+
+        # Local table which maps the name of a binry operator to the
+        # corresponding Python function.
+        op_to_python = {
+            "Add": "__add__",
+            "BitAnd": "__and__",
+            "LShift": "__lshift__",
+            "Mult": "__mul__",
+            "RShift": "__rshift__",
+            "Sub": "__sub__",
+        }
+        if op_name not in op_to_python:
+            raise CodeGenerationException(
+                node.lineno, node.col_offset,
+                f"Unknown binary operation {op_name}.")
+
+        self.visit(node.right)
+        rhs = self.inserter.get_operand()
+        self.visit(node.left)
+        lhs = self.inserter.get_operand()
+        if lhs.typ != rhs.typ:
```

Comment:
Python actually allows adding ints to floats. Do we just redefine semantics here? Or is this an oversight?

Reply:
No, we just guard for the case we not support (as a side note, should I change the error to "not supported yet"?) The main issue is we want to have implicit casting and just insert them in cases like float to int addition (`fptosi`, `extsi`, etc.). The issue with this is of course Python, because we have to be way more careful with implicit casts otherwise we can end up converting `u256` to `i8`, say :)